### PR TITLE
Improve error recovery for unterminated string literals.

### PR DESCRIPTION
**Pull request by A-thanasios**
#56

### CODE REVIEW

File: test/test-lex.cpp [hunk 1]
```diff
@@ -525,6 +525,72 @@ TEST(test_lex, lex_strings) {
...
+    padded_string input(u8"'separated" + string8(line_terminator) + u8"hello'");
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::string);
+    l.skip();
+    EXPECT_EQ(l.peek().type, token_type::end_of_file);
+
+    EXPECT_THAT(v.errors,
+                ElementsAre(ERROR_TYPE_FIELD(
+                    error_unclosed_string_literal, string_literal,
+                    offsets_matcher(&input, 0, 16 + line_terminator.size()))));
```
Comment (strager) [inline test/test-lex.cpp:539]: Could we use `input.size()` instead of `16 + line_terminator.size()` here?

File: test/test-lex.cpp [hunk 2]
```diff
@@ -525,6 +525,72 @@ TEST(test_lex, lex_strings) {
...
+    EXPECT_THAT(
+        v.errors,
+        ElementsAre(
+            ERROR_TYPE_FIELD(error_unclosed_string_literal, string_literal,
+                             offsets_matcher(&input, 0, 10)),
+            ERROR_TYPE_FIELD(
+                error_unclosed_string_literal, string_literal,
+                offsets_matcher(&input, 15 + 2 * line_terminator.size(),
+                                16 + 2 * line_terminator.size()))));
+  }
```
Comment (strager) [inline test/test-lex.cpp:564]: :+1: I like this test.

File: src/lex.cpp [hunk 3]
```diff
@@ -450,10 +450,34 @@ void lexer::parse_current_token() {
         }
 
       case '\n':
-      case '\r':
+      case '\r': {
+        char8* c_up_to_first_matching_quote = nullptr;
```
Comment (strager) [inline src/lex.cpp:454]: Nit: I'd name this variable `matching_quote`, and decrement it. (See my comment on lines 466-470.)

File: src/lex.cpp [hunk 4]
```diff
@@ -450,10 +450,34 @@ void lexer::parse_current_token() {
...
+        for (;;) {
+          if (*c_end == opening_quote) {
+            if (c_up_to_first_matching_quote) {
+              break;
+            }
+            c_up_to_first_matching_quote = ++c_end;
+          } else if (*c_end == '\r' || *c_end == '\n' ||
+                     (*c_end == '\0' && this->is_eof(c_end))) {
+            if (c_up_to_first_matching_quote) {
+              c = c_up_to_first_matching_quote;
```
Comment (strager) [inline src/lex.cpp:470]: (From comment on line 454)

This is what I mean by 'decrement':
```suggestion
            c_up_to_first_matching_quote = c_end;
            ++c_end;
          } else if (*c_end == '\r' || *c_end == '\n' ||
                     (*c_end == '\0' && this->is_eof(c_end))) {
            if (c_up_to_first_matching_quote) {
              c = c_up_to_first_matching_quote + 1;
```

File: src/lex.cpp [hunk 5]
```diff
@@ -450,10 +450,34 @@ void lexer::parse_current_token() {
         }
 
       case '\n':
-      case '\r':
+      case '\r': {
+        char8* c_up_to_first_matching_quote = nullptr;
+        char8* c_end = c;
```
Comment (strager) [inline src/lex.cpp:455]: Hmm, do you think this code would be better in its own function? In a separate function, better variable names might come more naturally.

**Review by strager — APPROVED** [review]

File: test/test-lex.cpp (hunk 2 above)
Comment (A-thanasios) [reply test/test-lex.cpp:563]: Thanks :hear_no_evil:

**Comment by A-thanasios** [comment]
Ready to merge

**Comment by strager** [comment]
It looks like some builds are failing on CI: https://github.com/quick-lint/quick-lint-js/runs/1535159043

Please fix the builds. (I think you want a `narrow_cast<int>(...)`.)

**Comment by strager** [comment]
Landed as commits b879f5138, 93e63d2c5, and 65e936aaa. Thanks!

**Pull request closed without merging**